File: crates/kuben-core/src/scan.rs
//! Image SBOMs and vulnerability scans (M4.6; plan §13.3, §18.1).
//!
//! Every built image is scanned by digest; the scan records the scanner and
//! the age of its vulnerability database, because a scan is only as fresh as
//! its feed and an unavailable feed is never "clean". An environment's gate
//! decides what a deployment may carry: nothing (off), a warning, or a
//! refusal of findings at or above a severity that no unexpired exception
//! covers, and optionally of images without a fresh scan. Provenance,
//! signatures and findings are separate questions; this module answers only
//! the last.

use std::collections::BTreeSet;

use serde::{Deserialize, Serialize};

/// Findings listed by id in a scan, at most.
pub const MAX_LISTED: usize = 40;
/// Longest exception a person may grant.
pub const MAX_EXCEPTION_SECS: i64 = 90 * 24 * 3600;
/// Default age after which a scan no longer counts.
pub const DEFAULT_MAX_AGE_SECS: u32 = 7 * 24 * 3600;

/// A finding's severity, as scanners report it.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Unknown,
    Low,
    Medium,
    High,
    Critical,
}

impl Severity {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Unknown => "unknown",
            Self::Low => "low",
            Self::Medium => "medium",
            Self::High => "high",
            Self::Critical => "critical",
        }
    }

    #[must_use]
    pub fn parse(s: &str) -> Option<Self> {
        Some(match s.to_ascii_lowercase().as_str() {
            "unknown" => Self::Unknown,
            "low" => Self::Low,
            "medium" => Self::Medium,
            "high" => Self::High,
            "critical" => Self::Critical,
            _ => return None,
        })
    }
}

/// Findings per severity.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Counts {
    #[serde(default)]
    pub critical: u32,
    #[serde(default)]
    pub high: u32,
    #[serde(default)]
    pub medium: u32,
    #[serde(default)]
    pub low: u32,
    #[serde(default)]
    pub unknown: u32,
}

impl Counts {
    /// Findings at `severity` or above.
    #[must_use]
    pub const fn at_least(&self, severity: Severity) -> u32 {
        let mut n = self.critical;
        if matches!(
            severity,
            Severity::High | Severity::Medium | Severity::Low | Severity::Unknown
        ) {
            n = n.saturating_add(self.high);
        }
        if matches!(severity, Severity::Medium | Severity::Low | Severity::Unknown) {
            n = n.saturating_add(self.medium);
        }
        if matches!(severity, Severity::Low | Severity::Unknown) {
            n = n.saturating_add(self.low);
        }
        if matches!(severity, Severity::Unknown) {
            n = n.saturating_add(self.unknown);
        }
        n
    }
}

/// Whether a scan could run.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ScanStatus {
    Ok,
    /// The scanner, its feed or the image was not available.
    Unavailable,
}

impl ScanStatus {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Ok => "ok",
            Self::Unavailable => "unavailable",
        }
    }
}

/// One finding by id: `(severity, id)`.
pub type Finding = (Severity, String);

/// What a scan container writes to its termination log (at most 4 KiB).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ScanReport {
    pub status: ScanStatus,
    #[serde(default)]
    pub scanner: String,
    /// When the vulnerability database was built (RFC 3339).
    #[serde(default)]
    pub db: Option<String>,
    #[serde(default)]
    pub counts: Counts,
    /// `SEVERITY:ID`, most severe first.
    #[serde(default)]
    pub findings: Vec<String>,
    #[serde(default)]
    pub detail: Option<String>,
}

impl ScanReport {
    /// A report of a scan that could not run.
    #[must_use]
    pub fn unavailable(detail: impl Into<String>) -> Self {
        Self {
            status: ScanStatus::Unavailable,
            scanner: String::new(),
            db: None,
            counts: Counts::default(),
            findings: Vec::new(),
            detail: Some(detail.into()),
        }
    }

    /// The listed findings that parse, at most [`MAX_LISTED`].
    #[must_use]
    pub fn listed(&self) -> Vec<Finding> {
        self.findings
            .iter()
            .filter_map(|f| {
                let (severity, id) = f.split_once(':')?;
                let id = id.trim();
                let valid = !id.is_empty()
                    && id.len() <= 64
                    && id
                        .bytes()
                        .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'-' | b'_' | b'.' | b':'));
                valid.then_some((Severity::parse(severity)?, id.to_owned()))
            })
            .take(MAX_LISTED)
            .collect()
    }
}

/// A recorded scan of one digest.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ScanSummary {
    pub status: ScanStatus,
    pub scanner: String,
    /// When the vulnerability database was built (Unix milliseconds).
    pub db_updated_at: Option<i64>,
    pub counts: Counts,
    pub findings: Vec<Finding>,
    /// When the scan ran (Unix milliseconds).
    pub scanned_at: i64,
}

/// What an environment's gate does.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum GateMode {
    #[default]
    Off,
    Warn,
    Block,
}

impl GateMode {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Off => "off",
            Self::Warn => "warn",
            Self::Block => "block",
        }
    }

    #[must_use]
    pub fn parse(s: &str) -> Option<Self> {
        Some(match s {
            "off" => Self::Off,
            "warn" => Self::Warn,
            "block" => Self::Block,
            _ => return None,
        })
    }
}

/// An environment's vulnerability gate.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScanGate {
    pub mode: GateMode,
    /// Findings at or above this severity count (`high` or `critical`).
    pub severity: Severity,
    /// An image without a fresh, successful scan counts as a finding.
    pub require_scan: bool,
    /// A scan older than this no longer counts.
    pub max_age_secs: u32,
}

impl Default for ScanGate {
    fn default() -> Self {
        Self::off()
    }
}

impl ScanGate {
    #[must_use]
    pub const fn off() -> Self {
        Self {
            mode: GateMode::Off,
            severity: Severity::Critical,
            require_scan: false,
            max_age_secs: DEFAULT_MAX_AGE_SECS,
        }
    }

    /// Production: known critical findings are refused.
    #[must_use]
    pub const fn production() -> Self {
        Self {
            mode: GateMode::Block,
            ..Self::off()
        }
    }

    #[must_use]
    pub const fn is_valid(&self) -> bool {
        matches!(self.severity, Severity::High | Severity::Critical)
            && self.max_age_secs >= 3600
            && self.max_age_secs <= 90 * 24 * 3600
    }

    /// Whether `self` refuses less than `current`.
    #[must_use]
    pub fn weakens(&self, current: &Self) -> bool {
        self.mode < current.mode
            || self.severity > current.severity
            || (current.require_scan && !self.require_scan)
            || self.max_age_secs > current.max_age_secs
    }
}

/// What the gate decided for a deployment.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum GateVerdict {
    Pass,
    Warn(Vec<String>),
    Block(Vec<String>),
}

/// Decide on the images `scans` (digest and its newest scan) under `gate`
/// at `now` (Unix milliseconds); `excepted` are the finding ids with an
/// unexpired exception.
#[must_use]
pub fn evaluate(
    gate: &ScanGate,
    scans: &[(String, Option<ScanSummary>)],
    excepted: &BTreeSet<String>,
    now: i64,
) -> GateVerdict {
    if gate.mode == GateMode::Off {
        return GateVerdict::Pass;
    }
    let mut reasons = Vec::new();
    for (digest, scan) in scans {
        let short = digest.get(..19).unwrap_or(digest);
        let fresh = scan.as_ref().filter(|s| {
            s.status == ScanStatus::Ok
                && now.saturating_sub(s.scanned_at) <= i64::from(gate.max_age_secs) * 1000
        });
        let Some(scan) = fresh else {
            if gate.require_scan {
                reasons.push(format!("{short}… has no fresh vulnerability scan"));
            }
            continue;
        };
        let counted = scan.counts.at_least(gate.severity);
        let covered = scan
            .findings
            .iter()
            .filter(|(severity, id)| *severity >= gate.severity && excepted.contains(id))
            .count();
        let open = counted.saturating_sub(u32::try_from(covered).unwrap_or(u32::MAX));
        if open > 0 {
            let ids: Vec<&str> = scan
                .findings
                .iter()
                .filter(|(severity, id)| *severity >= gate.severity && !excepted.contains(id))
                .map(|(_, id)| id.as_str())
                .take(5)
                .collect();
            reasons.push(format!(
                "{short}… has {open} {} or worse finding(s) without an exception{}",
                gate.severity.as_str(),
                if ids.is_empty() {
                    String::new()
                } else {
                    format!(": {}", ids.join(", "))
                }
            ));
        }
    }
    match (reasons.is_empty(), gate.mode) {
        (true, _) | (_, GateMode::Off) => GateVerdict::Pass,
        (false, GateMode::Warn) => GateVerdict::Warn(reasons),
        (false, GateMode::Block) => GateVerdict::Block(reasons),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const DIGEST: &str = "sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
    const NOW: i64 = 1_800_000_000_000;

    fn scan(counts: Counts, findings: &[(Severity, &str)], age_secs: i64) -> ScanSummary {
        ScanSummary {
            status: ScanStatus::Ok,
            scanner: "trivy 0.74.0".into(),
            db_updated_at: Some(NOW - 3_600_000),
            counts,
            findings: findings.iter().map(|(s, id)| (*s, (*id).to_owned())).collect(),
            scanned_at: NOW - age_secs * 1000,
        }
    }

    fn one(s: Option<ScanSummary>) -> Vec<(String, Option<ScanSummary>)> {
        vec![(DIGEST.to_owned(), s)]
    }

    #[test]
    fn reports_list_only_well_formed_findings() {
        let report: ScanReport = serde_json::from_str(
            r#"{"status":"ok","scanner":"trivy 0.74.0","db":"2026-09-17T00:00:00Z",
                "counts":{"critical":1,"high":2},
                "findings":["CRITICAL:CVE-2026-0001","HIGH:GHSA-abcd-efgh","BOGUS:CVE-1","HIGH:","HIGH:a b"]}"#,
        )
        .expect("report");
        assert_eq!(
            report.listed(),
            [
                (Severity::Critical, "CVE-2026-0001".to_owned()),
                (Severity::High, "GHSA-abcd-efgh".to_owned())
            ]
        );
        assert_eq!(report.counts.at_least(Severity::High), 3);
        assert_eq!(report.counts.at_least(Severity::Critical), 1);
        let down: ScanReport =
            serde_json::from_str(r#"{"status":"unavailable","detail":"no feed"}"#).expect("r");
        assert_eq!(down, ScanReport::unavailable("no feed"));
    }

    #[test]
    fn open_findings_above_the_severity_are_refused() {
        let gate = ScanGate::production();
        let critical = Counts {
            critical: 1,
            high: 4,
            ..Counts::default()
        };
        let findings = [(Severity::Critical, "CVE-1"), (Severity::High, "CVE-2")];
        let verdict = evaluate(
            &gate,
            &one(Some(scan(critical, &findings, 60))),
            &BTreeSet::new(),
            NOW,
        );
        assert!(
            matches!(&verdict, GateVerdict::Block(r) if r[0].contains("1 critical") && r[0].contains("CVE-1"))
        );
        let excepted = BTreeSet::from(["CVE-1".to_owned()]);
        assert_eq!(
            evaluate(&gate, &one(Some(scan(critical, &findings, 60))), &excepted, NOW),
            GateVerdict::Pass,
            "the critical finding has an exception; high ones do not count"
        );
        let high = ScanGate {
            severity: Severity::High,
            ..gate
        };
        assert!(matches!(
            evaluate(&high, &one(Some(scan(critical, &findings, 60))), &excepted, NOW),
            GateVerdict::Block(r) if r[0].contains("4 high")
        ));
        let warn = ScanGate {
            mode: GateMode::Warn,
            ..gate
        };
        assert!(matches!(
            evaluate(
                &warn,
                &one(Some(scan(critical, &findings, 60))),
                &BTreeSet::new(),
                NOW
            ),
            GateVerdict::Warn(_)
        ));
        assert_eq!(
            evaluate(
                &ScanGate::off(),
                &one(Some(scan(critical, &findings, 60))),
                &BTreeSet::new(),
                NOW
            ),
            GateVerdict::Pass
        );
    }

    #[test]
    fn unlisted_findings_cannot_be_excepted() {
        // Two critical findings, only one of them listed and excepted.
        let counts = Counts {
            critical: 2,
            ..Counts::default()
        };
        let excepted = BTreeSet::from(["CVE-1".to_owned()]);
        let s = scan(counts, &[(Severity::Critical, "CVE-1")], 60);
        assert!(matches!(
            evaluate(&ScanGate::production(), &one(Some(s)), &excepted, NOW),
            GateVerdict::Block(r) if r[0].contains("1 critical")
        ));
    }

    #[test]
    fn missing_stale_or_unavailable_scans_count_only_when_required() {
        let gate = ScanGate::production();
        let clean = scan(Counts::default(), &[], 60);
        let stale = scan(Counts::default(), &[], i64::from(gate.max_age_secs) + 1);
        let unavailable = ScanSummary {
            status: ScanStatus::Unavailable,
            ..clean.clone()
        };
        for s in [None, Some(stale.clone()), Some(unavailable.clone())] {
            assert_eq!(
                evaluate(&gate, &one(s.clone()), &BTreeSet::new(), NOW),
                GateVerdict::Pass
            );
            let strict = ScanGate {
                require_scan: true,
                ..gate
            };
            assert!(matches!(
                evaluate(&strict, &one(s), &BTreeSet::new(), NOW),
                GateVerdict::Block(r) if r[0].contains("no fresh vulnerability scan")
            ));
        }
        // A stale scan's findings do not count either: rescans refresh them.
        let stale_critical = ScanSummary {
            counts: Counts {
                critical: 3,
                ..Counts::default()
            },
            ..stale
        };
        assert_eq!(
            evaluate(&gate, &one(Some(stale_critical)), &BTreeSet::new(), NOW),
            GateVerdict::Pass
        );
    }

    #[test]
    fn gates_weaken_in_every_direction() {
        let strict = ScanGate {
            mode: GateMode::Block,
            severity: Severity::High,
            require_scan: true,
            max_age_secs: 86_400,
        };
        assert!(!strict.weakens(&strict));
        for weaker in [
            ScanGate {
                mode: GateMode::Warn,
                ..strict
            },
            ScanGate {
                severity: Severity::Critical,
                ..strict
            },
            ScanGate {
                require_scan: false,
                ..strict
            },
            ScanGate {
                max_age_secs: 172_800,
                ..strict
            },
        ] {
            assert!(weaker.weakens(&strict), "{weaker:?}");
            assert!(!strict.weakens(&weaker));
        }
        assert!(strict.is_valid());
        assert!(
            !ScanGate {
                severity: Severity::Low,
                ..strict
            }
            .is_valid()
        );
        assert!(
            !ScanGate {
                max_age_secs: 10,
                ..strict
            }
            .is_valid()
        );
        assert_eq!(GateMode::parse("block"), Some(GateMode::Block));
        assert_eq!(Severity::parse("HIGH"), Some(Severity::High));
    }
}

File: apps/kuben/internal/core/scan/scan.go
// Package scan has image SBOMs and vulnerability scans (M4.6; plan §13.3,
// §18.1). It replaces the Rust module kuben-core/src/scan.rs.
//
// Every built image is scanned by digest; the scan records the scanner and
// the age of its vulnerability database, because a scan is only as fresh as
// its feed and an unavailable feed is never "clean". An environment's gate
// decides what a deployment may carry: nothing (off), a warning, or a
// refusal of findings at or above a severity that no unexpired exception
// covers, and optionally of images without a fresh scan. Provenance,
// signatures and findings are separate questions; this package answers only
// the last.
package scan

import (
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"unicode/utf8"

	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/ascii"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/clock"

	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/kerrors"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/opt"
)

const (
	// MaxListed is how many findings a scan lists by id, at most.
	MaxListed = 40
	// MaxExceptionSecs is the longest exception a person may grant.
	MaxExceptionSecs int64 = 90 * 24 * 3600
	// DefaultMaxAgeSecs is the default age after which a scan no longer counts.
	DefaultMaxAgeSecs uint32 = 7 * 24 * 3600

	minMaxAgeSecs uint32 = 3600
	maxMaxAgeSecs uint32 = 90 * 24 * 3600
	// shortDigest is how much of a digest a reason shows: "sha256:" and
	// twelve hex digits.
	shortDigest = 19
	// maxFindingID is the longest finding id that is listed.
	maxFindingID = 64
	// reasonIDs is how many open finding ids a reason names.
	reasonIDs = 5
)

// Severity is a finding's severity, as scanners report it. Severities are
// ordered by [Severity.Rank], never by their strings.
type Severity string

// The severities, from least to most severe.
const (
	SeverityUnknown  Severity = "unknown"
	SeverityLow      Severity = "low"
	SeverityMedium   Severity = "medium"
	SeverityHigh     Severity = "high"
	SeverityCritical Severity = "critical"
)

// ParseSeverity reads a severity name in any ASCII case, as scanners write
// them ("HIGH", "high").
func ParseSeverity(s string) (Severity, error) {
	switch v := Severity(ascii.Lower(s)); v {
	case SeverityUnknown, SeverityLow, SeverityMedium, SeverityHigh, SeverityCritical:
		return v, nil
	}
	return "", kerrors.New(kerrors.Validation, "unknown severity `%s`", s)
}

// Rank is the severity's position in the order unknown < low < medium <
// high < critical; -1 for a value that is not a severity.
func (s Severity) Rank() int {
	switch s {
	case SeverityUnknown:
		return 0
	case SeverityLow:
		return 1
	case SeverityMedium:
		return 2
	case SeverityHigh:
		return 3
	case SeverityCritical:
		return 4
	}
	return -1
}

// Compare is negative when s is less severe than other, zero when they are
// equally severe and positive when s is more severe.
func (s Severity) Compare(other Severity) int { return s.Rank() - other.Rank() }

func (s Severity) String() string { return string(s) }

// UnmarshalJSON accepts only the exact wire names.
func (s *Severity) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err != nil {
		return err
	}
	v := Severity(text)
	if v.Rank() < 0 {
		return kerrors.New(kerrors.Validation, "unknown severity `%s`", text)
	}
	*s = v
	return nil
}

// Counts are the findings per severity.
type Counts struct {
	Critical uint32 `json:"critical"`
	High     uint32 `json:"high"`
	Medium   uint32 `json:"medium"`
	Low      uint32 `json:"low"`
	Unknown  uint32 `json:"unknown"`
}

// AtLeast is the number of findings at severity or above, saturating. A
// value that is not a severity counts critical findings only.
func (c Counts) AtLeast(severity Severity) uint32 {
	n := c.Critical
	rank := severity.Rank()
	if rank < 0 {
		return n
	}
	for _, level := range []struct {
		rank  int
		count uint32
	}{
		{SeverityHigh.Rank(), c.High},
		{SeverityMedium.Rank(), c.Medium},
		{SeverityLow.Rank(), c.Low},
		{SeverityUnknown.Rank(), c.Unknown},
	} {
		if rank <= level.rank {
			n = saturatingAddU32(n, level.count)
		}
	}
	return n
}

// Status says whether a scan could run.
type Status string

// The scan statuses.
const (
	StatusOK Status = "ok"
	// StatusUnavailable: the scanner, its feed or the image was not available.
	StatusUnavailable Status = "unavailable"
)

// ParseStatus reads a scan status.
func ParseStatus(s string) (Status, error) {
	switch v := Status(s); v {
	case StatusOK, StatusUnavailable:
		return v, nil
	}
	return "", kerrors.New(kerrors.Validation, "unknown scan status `%s`", s)
}

func (s Status) String() string { return string(s) }

// UnmarshalJSON accepts only the exact wire names.
func (s *Status) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err != nil {
		return err
	}
	v, err := ParseStatus(text)
	if err != nil {
		return err
	}
	*s = v
	return nil
}

// Finding is one finding by id.
type Finding struct {
	Severity Severity
	ID       string
}

// Report is what a scan container writes to its termination log (at most
// 4 KiB).
type Report struct {
	Status  Status `json:"status"`
	Scanner string `json:"scanner"`
	// DB is when the vulnerability database was built (RFC 3339).
	DB     opt.Val[string] `json:"db"`
	Counts Counts          `json:"counts"`
	// Findings are `SEVERITY:ID`, most severe first.
	Findings []string        `json:"findings"`
	Detail   opt.Val[string] `json:"detail"`
}

// Unavailable is the report of a scan that could not run.
func Unavailable(detail string) Report {
	return Report{Status: StatusUnavailable, Findings: []string{}, Detail: opt.Some(detail)}
}

// MarshalJSON writes no findings as `[]`, never as null.
func (r Report) MarshalJSON() ([]byte, error) {
	type wire Report
	w := wire(r)
	if w.Findings == nil {
		w.Findings = []string{}
	}
	return json.Marshal(w)
}

// UnmarshalJSON requires the status; everything else defaults to empty.
func (r *Report) UnmarshalJSON(data []byte) error {
	type wire Report
	var w wire
	if err := json.Unmarshal(data, &w); err != nil {
		return err
	}
	if w.Status == "" {
		return kerrors.New(kerrors.Validation, "missing field `status`")
	}
	if w.Findings == nil {
		w.Findings = []string{}
	}
	*r = Report(w)
	return nil
}

// Listed are the listed findings that parse, at most [MaxListed].
func (r Report) Listed() []Finding {
	listed := []Finding{}
	for _, f := range r.Findings {
		if len(listed) == MaxListed {
			break
		}
		name, id, found := strings.Cut(f, ":")
		if !found {
			continue
		}
		id = strings.TrimSpace(id)
		if !validFindingID(id) {
			continue
		}
		severity, err := ParseSeverity(name)
		if err != nil {
			continue
		}
		listed = append(listed, Finding{Severity: severity, ID: id})
	}
	return listed
}

func validFindingID(id string) bool {
	if id == "" || len(id) > maxFindingID {
		return false
	}
	for i := range len(id) {
		b := id[i]
		alnum := b >= '0' && b <= '9' || b >= 'a' && b <= 'z' || b >= 'A' && b <= 'Z'
		if !alnum && b != '-' && b != '_' && b != '.' && b != ':' {
			return false
		}
	}
	return true
}

// Summary is a recorded scan of one digest.
type Summary struct {
	Status  Status
	Scanner string
	// DBUpdatedAt is when the vulnerability database was built (unix ms).
	DBUpdatedAt opt.Val[int64]
	Counts      Counts
	Findings    []Finding
	// ScannedAt is when the scan ran (unix ms).
	ScannedAt int64
}

// GateMode is what an environment's gate does. Modes are ordered by
// [GateMode.Rank]: off < warn < block.
type GateMode string

// The gate modes, from weakest to strictest.
const (
	ModeOff   GateMode = "off"
	ModeWarn  GateMode = "warn"
	ModeBlock GateMode = "block"
)

// ParseGateMode reads a gate mode.
func ParseGateMode(s string) (GateMode, error) {
	switch v := GateMode(s); v {
	case ModeOff, ModeWarn, ModeBlock:
		return v, nil
	}
	return "", kerrors.New(kerrors.Validation, "unknown gate mode `%s`", s)
}

// Rank is the mode's strictness; -1 for a value that is not a mode.
func (m GateMode) Rank() int {
	switch m {
	case ModeOff:
		return 0
	case ModeWarn:
		return 1
	case ModeBlock:
		return 2
	}
	return -1
}

func (m GateMode) String() string { return string(m) }

// UnmarshalJSON accepts only the exact wire names.
func (m *GateMode) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err != nil {
		return err
	}
	v, err := ParseGateMode(text)
	if err != nil {
		return err
	}
	*m = v
	return nil
}

// Gate is an environment's vulnerability gate. The zero Gate is not a valid
// gate; [Off] is the default one.
type Gate struct {
	Mode GateMode `json:"mode"`
	// Severity: findings at or above it count (`high` or `critical`).
	Severity Severity `json:"severity"`
	// RequireScan: an image without a fresh, successful scan counts as a
	// finding.
	RequireScan bool `json:"requireScan"`
	// MaxAgeSecs: a scan older than this no longer counts.
	MaxAgeSecs uint32 `json:"maxAgeSecs"`
}

// Off is the default gate: nothing is checked.
func Off() Gate {
	return Gate{Mode: ModeOff, Severity: SeverityCritical, RequireScan: false, MaxAgeSecs: DefaultMaxAgeSecs}
}

// Production is the gate of a production environment: known critical
// findings are refused.
func Production() Gate {
	g := Off()
	g.Mode = ModeBlock
	return g
}

// UnmarshalJSON requires every field, as the stored and the API form do.
func (g *Gate) UnmarshalJSON(data []byte) error {
	var w struct {
		Mode        opt.Val[GateMode] `json:"mode"`
		Severity    opt.Val[Severity] `json:"severity"`
		RequireScan opt.Val[bool]     `json:"requireScan"`
		MaxAgeSecs  opt.Val[uint32]   `json:"maxAgeSecs"`
	}
	if err := json.Unmarshal(data, &w); err != nil {
		return err
	}
	mode, hasMode := w.Mode.Get()
	severity, hasSeverity := w.Severity.Get()
	requireScan, hasRequireScan := w.RequireScan.Get()
	maxAge, hasMaxAge := w.MaxAgeSecs.Get()
	for _, field := range []struct {
		name    string
		present bool
	}{{"mode", hasMode}, {"severity", hasSeverity}, {"requireScan", hasRequireScan}, {"maxAgeSecs", hasMaxAge}} {
		if !field.present {
			return kerrors.New(kerrors.Validation, "missing field `%s`", field.name)
		}
	}
	*g = Gate{Mode: mode, Severity: severity, RequireScan: requireScan, MaxAgeSecs: maxAge}
	return nil
}

// Valid reports whether the gate may be stored: a known mode, a severity of
// `high` or `critical`, and a maximum age of an hour to 90 days.
func (g Gate) Valid() bool {
	return g.Mode.Rank() >= 0 &&
		(g.Severity == SeverityHigh || g.Severity == SeverityCritical) &&
		g.MaxAgeSecs >= minMaxAgeSecs && g.MaxAgeSecs <= maxMaxAgeSecs
}

// Weakens reports whether g refuses less than current.
func (g Gate) Weakens(current Gate) bool {
	return g.Mode.Rank() < current.Mode.Rank() ||
		g.Severity.Compare(current.Severity) > 0 ||
		(current.RequireScan && !g.RequireScan) ||
		g.MaxAgeSecs > current.MaxAgeSecs
}

// Verdict is what the gate decided for a deployment.
//
//sumtype:decl
type Verdict interface{ verdict() }

// Pass means the deployment may go ahead.
type Pass struct{}

// Warn means the deployment may go ahead, with these reasons shown.
type Warn struct{ Reasons []string }

// Block means the deployment is refused for these reasons.
type Block struct{ Reasons []string }

func (Pass) verdict()  {}
func (Warn) verdict()  {}
func (Block) verdict() {}

// Scanned is an image of a deployment: its digest and its newest scan, if
// it has one.
type Scanned struct {
	Digest string
	Scan   opt.Val[Summary]
}

// Evaluate decides on the images scans under gate at now (unix ms); excepted
// are the finding ids with an unexpired exception. A gate whose mode is not
// a known mode checks nothing, like [ModeOff].
func Evaluate(gate Gate, scans []Scanned, excepted map[string]struct{}, now int64) Verdict {
	if gate.Mode != ModeWarn && gate.Mode != ModeBlock {
		return Pass{}
	}
	var reasons []string
	for _, image := range scans {
		short := shorten(image.Digest)
		scan, ok := image.Scan.Get()
		fresh := ok && scan.Status == StatusOK &&
			clock.SaturatingSub(now, scan.ScannedAt) <= int64(gate.MaxAgeSecs)*1000
		if !fresh {
			if gate.RequireScan {
				reasons = append(reasons, short+"… has no fresh vulnerability scan")
			}
			continue
		}
		if reason, open := openFindings(gate, scan, excepted); open {
			reasons = append(reasons, short+reason)
		}
	}
	switch {
	case len(reasons) == 0:
		return Pass{}
	case gate.Mode == ModeWarn:
		return Warn{Reasons: reasons}
	default:
		return Block{Reasons: reasons}
	}
}

// openFindings is the reason (without the digest) a fresh scan counts
// against the gate, if it does. Findings that are counted but not listed
// cannot be excepted.
func openFindings(gate Gate, scan Summary, excepted map[string]struct{}) (string, bool) {
	counted := scan.Counts.AtLeast(gate.Severity)
	var covered uint32
	var ids []string
	for _, f := range scan.Findings {
		if f.Severity.Compare(gate.Severity) < 0 {
			continue
		}
		if _, ok := excepted[f.ID]; ok {
			if covered < math.MaxUint32 {
				covered++
			}
		} else if len(ids) < reasonIDs {
			ids = append(ids, f.ID)
		}
	}
	if covered >= counted {
		return "", false
	}
	reason := fmt.Sprintf("… has %d %s or worse finding(s) without an exception", counted-covered, gate.Severity)
	if len(ids) > 0 {
		reason += ": " + strings.Join(ids, ", ")
	}
	return reason, true
}

// shorten is the first [shortDigest] bytes of digest, or all of it when it
// is shorter or that would split a character.
func shorten(digest string) string {
	if len(digest) < shortDigest {
		return digest
	}
	if len(digest) > shortDigest && !utf8.RuneStart(digest[shortDigest]) {
		return digest
	}
	return digest[:shortDigest]
}

func saturatingAddU32(a, b uint32) uint32 {
	if a > math.MaxUint32-b {
		return math.MaxUint32
	}
	return a + b
}
